core/vm/ipgraph_with_kind.go

```diff
@@ -0,0 +1,365 @@
+	parentIpId := common.BytesToAddress(input[32:64])
+	royaltyPolicyKind := new(big.Int).SetBytes(getData(input, 64, 32))
+	royalty := new(big.Int).SetBytes(getData(input, 96, 32))
+	slot := crypto.Keccak256Hash(ipId.Bytes(), parentIpId.Bytes(), royaltyPolicyKind.Bytes()).Big()
+	log.Info("setRoyalty", "ipId", "ipGraphAddress", ipGraphAddress, ipId, "parentIpId", parentIpId,
+		"royaltyPolicyKind", royaltyPolicyKind, "royalty", royalty, "slot", slot)
+	evm.StateDB.SetState(ipGraphAddress, common.BigToHash(slot), common.BigToHash(royalty))
+
+	return nil, nil
+}
+
+func (c *ipGraphWithPolicyKind) getRoyaltyStackWithKind(input []byte, evm *EVM, ipGraphAddress common.Address) ([]byte, error) {
+	log.Info("getRoyaltyStackKind", "ipGraphAddress", ipGraphAddress, "input", input)
+	totalRoyalty := big.NewInt(0)
+	if len(input) < 32 {
+		return nil, fmt.Errorf("input too short for getRoyaltyStack")
+	}
+	ipId := common.BytesToAddress(input[0:32])
+	royaltyPolicyKind := new(big.Int).SetBytes(getData(input, 32, 32))
+	if royaltyPolicyKind.Cmp(royaltyPolicyKindLAP) == 0 {
+		totalRoyalty = c.getRoyaltyStackLap(ipId, evm, ipGraphAddress)
+	} else if royaltyPolicyKind.Cmp(royaltyPolicyKindLRP) == 0 {
+		totalRoyalty = c.getRoyaltyStackLrp(ipId, evm, ipGraphAddress)
+	} else {
+		return nil, fmt.Errorf("unknown royalty policy kind")
+	}
+	log.Info("getRoyaltyStackKind", "ipId", ipId, "ipGraphAddress", ipGraphAddress, "royaltyPolicyKind", royaltyPolicyKind, "totalRoyalty", totalRoyalty)
+	return common.BigToHash(totalRoyalty).Bytes(), nil
+}
+
+func (c *ipGraphWithPolicyKind) getRoyaltyWithKind(input []byte, evm *EVM, ipGraphAddress common.Address) ([]byte, error) {
+	log.Info("getRoyaltyWithKind", "ipGraphAddress", ipGraphAddress, "input", input)
+	if len(input) < 64 {
+		return nil, fmt.Errorf("input too short for getRoyalty")
+	}
+	ipId := common.BytesToAddress(input[0:32])
+	ancestorIpId := common.BytesToAddress(input[32:64])
+	royaltyPolicyKind := new(big.Int).SetBytes(getData(input, 64, 32))
+	totalRoyalty := big.NewInt(0)
+	if royaltyPolicyKind.Cmp(royaltyPolicyKindLAP) == 0 {
+		totalRoyalty = c.getRoyaltyLap(ipId, ancestorIpId, evm, ipGraphAddress)
+	} else if royaltyPolicyKind.Cmp(royaltyPolicyKindLRP) == 0 {
+		totalRoyalty = c.getRoyaltyLrp(ipId, ancestorIpId, evm, ipGraphAddress)
+	} else {
+		return nil, fmt.Errorf("unknown royalty policy kind")
+	}
+
+	log.Info("getRoyaltyWithKind", "ipId", ipId, "ancestorIpId", ancestorIpId, "ipGraphAddress", ipGraphAddress, "royaltyPolicyKind", royaltyPolicyKind, "totalRoyalty", totalRoyalty)
+	return common.BigToHash(totalRoyalty).Bytes(), nil
+}
+
+func (c *ipGraphWithPolicyKind) getRoyaltyStackLap(ipId common.Address, evm *EVM, ipGraphAddress common.Address) *big.Int {
+	log.Info("getRoyaltyStackLap", "ipGraphAddress", ipGraphAddress, "IP ID", ipId)
+	ancestors := make(map[common.Address]struct{})
+	totalRoyalty := big.NewInt(0)
+	var stack []common.Address
+	stack = append(stack, ipId)
+	for len(stack) > 0 {
```

Comment:
We need to enforce a max len(stack) in the precompile, or risk a chain DOS
Even if there is a limit in the contracts, we need a reasonable max here.

Reply:
The limitation is currently defined the protocol contract, and new ipa can only be added to the precompiled ipgraph through the protocol contract due to ACL. it should be safe for dos.